core/src/main/java/com/linecorp/armeria/server/HttpHeaderUtil.java

```diff
@@ -141,32 +147,54 @@ private static InetAddress createInetAddress(String address) throws UnknownHostE
+        final InetSocketAddress inetSocketAddress;
+        try {
+            if (addr.charAt(0) == '[') {
+                inetSocketAddress = createInetSocketAddress(addr, ']');
+            } else {
+                inetSocketAddress = createInetSocketAddress(addr, ':');
+            }
+            // Log debug message if it is an invalid address.
+            if (inetSocketAddress == null) {
+                logger.debug("Failed to parse an address: {}", address);
+            }
+            return inetSocketAddress;
+        } catch (NumberFormatException e) {
+            logger.debug("Fail to parse a port from an address: {}", address, e);
+            return null;
         }
+    }
 
-        if (addr.charAt(0) == '[') {
-            final int delim = addr.indexOf(']');
-            if (delim > 0) {
-                // Remove the port part. e.g. [2001:db8:cafe::17]:4711 => [2001:db8:cafe::17]
-                final String withoutPort = addr.substring(0, delim + 1);
-                if (NetUtil.isValidIpV6Address(withoutPort)) {
-                    return InetAddress.getByAddress(NetUtil.createByteArrayFromIpAddressString(withoutPort));
-                }
-            }
+    @Nullable
+    private static InetSocketAddress createInetSocketAddress(String address, char delimiter)
+            throws UnknownHostException {
+        final int delim = address.indexOf(delimiter);
+        // Split addr into ip and port. e.g. [2001:db8:cafe::17]:4711 => [2001:db8:cafe::17], 4711
+        final String ip;
+        final int port;
+        if (delim == -1) {
+            ip = address;
+            port = 0;
         } else {
-            final int delim = addr.indexOf(':');
-            if (delim > 0) {
-                // Remove the port part. e.g. 192.168.1.1:4711 => 192.168.1.1
-                final String withoutPort = addr.substring(0, delim);
-                if (NetUtil.isValidIpV4Address(withoutPort)) {
-                    return InetAddress.getByAddress(NetUtil.createByteArrayFromIpAddressString(withoutPort));
+            if (delimiter == ']') {
+                ip = address.substring(0, delim + 1);
+                if (delim + 1 == address.length()) {
+                    port = 0;
+                } else {
+                    // skip ':'
+                    port = Integer.parseInt(address.substring(delim + 2));
+                }
+            } else {
+                ip = address.substring(0, delim);
+                if (delim + 1 == address.length()) {
+                    port = 0;
+                } else {
+                    port = Integer.parseInt(address.substring(delim + 1));
                 }
             }
         }
-        // Skip if it is an invalid address.
-        logger.debug("Failed to parse an address: {}", address);
-        return null;
+        return new InetSocketAddress(
+                InetAddress.getByAddress(NetUtil.createByteArrayFromIpAddressString(ip)), port);
```

Comment:
You need to handle the case where `createByteArrayFromIpAddressString()` returns `null`.